### DISCUSSION TITLE
What is the proper way to apply the update "gracefully"?

### DISCUSSION THREAD

**Discussion by Eoic** [dennisvang/tufup#96] [Q&A] [answered]
Ideally, I want to:

1. Check for updates and download if one is available.
2. Show update download progress to the user and apply the update without any further user interaction.
3. Re-open the application when the update is applied.

How do I achieve this? Does `tufup` provide any hooks or different API methods that I could use to implement such update process?

---

In the code below I check for and apply updates, if any were found. I run this code in a separate thread, so that it wouldn't block the UI and show download progress to the user.

```python
def apply_update(self):
    if not self.updater_client.check_for_updates():
        return

    self.updater_client.download_and_apply_update(
        purge_dst_dir=False,
        skip_confirmation=True,
        exclude_from_purge=None,
        progress_hook=lambda **download_args: self.update_progress_signal.emit(download_args),
    )
```

However, when the update is downloaded, it opens up a console window, but does not close the application. The output at this moment is shown below. Also, it cannot copy files when the application is running and I have to close the application manually for the update to succeed. 

```
Waiting 2 seconds... Retrying...
            Newer                 12.6 m        app.exe
2023/12/15 23:03:07 ERROR 32 (0x00000020) Copying File C:\Users\<USER>\AppData\Local\Temp\tufup\app.exe
The process cannot access the file because it is being used by another process.
Waiting 2 seconds... Retrying...
            Newer                 12.6 m        app.exe
```

**Comment by dennisvang** [comment, marked as answer]
>     1. Check for updates and download if one is available.

Basically do what you're doing in your example above.
 
>     2. Show update download progress to the user and apply the update without any further user interaction.

The `skip_confirmation=False` takes care fo the user interaction in the console. If you also want to prevent the console window from appearing, have a look at the `process_creation_flags` arg, as described in #90. 

That is, assuming you're on Windows.

>     3. Re-open the application when the update is applied.

There's a `batch_template` hook that allows you to provide a custom batch script. Have a look at [this discussion][1] for details.

On macOS the app is already restarted out-of-the-box. See [source][2]

>... I run this code in a separate thread, ...

Tufup was *not* designed with multi-threading in mind.

It can probably be done, but you'd have to jump through some hoops:

Off the top of my head, I think you could probably run the [`_download_updates()` step][4] in a separate thread, as long as you run the [`_apply_updates()` step][5] in your main thread. Tufup [checks][3] if the files for new updates have already been downloaded, so the update would then be applied on the next run of your app.

However, use *at your own risk*: this is untested and not officially supported.

I'm also not quite sure if, and how, the underlying `python-tuf` package handles multi-threading (see e.g. [8], [7], [6]).

>... However, when the update is downloaded, it opens up a console window, but does not close the application. ... Also, it cannot copy files when the application is running ...

That's probably all due to the multi-threading. The [`_apply_updates()` step][5] must run on the main thread.


[1]: https://github.com/dennisvang/tufup-example/issues/12#issuecomment-1398116894
[2]: https://github.com/dennisvang/tufup/blob/ff8a60db854f628aee6abb73de59c59e3d2d8d22/src/tufup/utils/platform_specific.py#L257
[3]: https://github.com/dennisvang/tufup/blob/ff8a60db854f628aee6abb73de59c59e3d2d8d22/src/tufup/client.py#L217
[4]: https://github.com/dennisvang/tufup/blob/ff8a60db854f628aee6abb73de59c59e3d2d8d22/src/tufup/client.py#L214
[5]: https://github.com/dennisvang/tufup/blob/ff8a60db854f628aee6abb73de59c59e3d2d8d22/src/tufup/client.py#L230
[6]: https://github.com/theupdateframework/python-tuf/commit/b913d88ad46ad4e97f24449443b96beb99546585
[7]: https://github.com/theupdateframework/python-tuf/commit/c3a358ed17723e5b8bd367c0fb352a33aa0dc62d
[8]: https://github.com/theupdateframework/python-tuf/issues/1158
